src/consensus/event_accumulator.rs

```diff
@@ -80,194 +90,168 @@ impl VoteStatuses {
+            .elder_ids()
+            .filter(|id| self.vote_statuses.is_unresponsive(id.name()))
+            .copied()
             .collect()
     }
 }
 
+// Accumulation state of a single event in the event accumulator.
 #[derive(Default, Debug, PartialEq, Eq, PartialOrd, Ord, Hash, Clone, Serialize, Deserialize)]
-pub struct AccumulatingProof {
-    parsec_proofs: ProofSet,
-    sig_shares: BTreeMap<PublicId, bls::SignatureShare>,
+struct State {
+    voters: BTreeSet<XorName>,
+    shares: BTreeMap<usize, bls::SignatureShare>,
 }
 
-impl AccumulatingProof {
-    #[cfg(test)]
-    pub fn from_proof_set(parsec_proofs: ProofSet) -> Self {
-        Self {
-            parsec_proofs,
-            sig_shares: Default::default(),
-        }
-    }
-
-    /// Return false if share or proof is replaced
-    pub fn add_proof(&mut self, proof: Proof, sig_share: Option<bls::SignatureShare>) -> bool {
-        let new_share = sig_share.map_or(true, |share| {
-            self.sig_shares.insert(proof.pub_id, share).is_none()
-        });
+impl State {
+    fn new(voter_name: XorName, voter_index: usize, signature_share: bls::SignatureShare) -> Self {
+        let voters = iter::once(voter_name).collect();
+        let shares = iter::once((voter_index, signature_share)).collect();
 
-        let new_proof = self.parsec_proofs.add_proof(proof);
-        new_share && new_proof
+        Self { voters, shares }
     }
 
-    pub fn parsec_proof_set(&self) -> &ProofSet {
-        &self.parsec_proofs
-    }
+    fn add(
+        &mut self,
+        voter_name: XorName,
+        voter_index: usize,
+        signature_share: bls::SignatureShare,
+    ) -> Result<(), AccumulatingError> {
+        if self.shares.insert(voter_index, signature_share).is_some() {
+            return Err(AccumulatingError::ReplacedAlreadyInserted);
+        }
 
-    pub fn into_sig_shares(self) -> BTreeMap<PublicId, bls::SignatureShare> {
-        self.sig_shares
+        if !self.voters.insert(voter_name) {
+            return Err(AccumulatingError::ReplacedAlreadyInserted);
+        }
+
+        Ok(())
     }
 
-    pub fn check_and_combine_signatures(
-        self,
-        elder_info: &EldersInfo,
-        pk_set: &bls::PublicKeySet,
-        signed_bytes: &[u8],
-    ) -> Option<bls::Signature> {
-        let fr_and_shares = elder_info
-            .elder_ids()
-            .enumerate()
-            .filter_map(|(index, pub_id)| {
-                self.sig_shares
-                    .get(pub_id)
-                    .map(|sig_share| (index, sig_share))
-            })
-            .filter(|&(index, sig_share)| {
-                pk_set
-                    .public_key_share(index)
-                    .verify(sig_share, signed_bytes)
-            });
-        pk_set.combine_signatures(fr_and_shares).ok()
+    fn has_enough_votes(&self, threshold: usize, elders_info: &EldersInfo) -> bool {
+        self.shares.len() > threshold && elders_info.is_quorum(&self.voters)
```

Comment:
Yeah. I think eventually we want to check only `shares.len() > threshold`. But checking the voters too for now because the threshold and the quorum are currently different (1/3 vs 2/3). But that's going to change with the DKG replacement, right?